File: src/scripts/sync_to_cloud.py
#!/usr/bin/env python3
import sys, os, json, subprocess
from concurrent.futures import ThreadPoolExecutor, as_completed

ENV_ID = 'cloudbase-d7gl3kh5vf6b71edc'
TCB = '/Users/zhigangwen/.nvm/versions/node/v24.15.0/bin/tcb'
SCRIPTS_DIR = os.path.dirname(os.path.abspath(__file__))
WORKERS = 10  # 并发数

def upsert_record(collection, record):
    cmd = [{"TableName": collection, "CommandType": "COMMAND", "Command": json.dumps({
        "findAndModify": collection,
        "query": {"_id": record['_id']},
        "update": {"$set": record},
        "upsert": True,
        "new": True
    })}]
    result = subprocess.run(
        [TCB, 'db', 'nosql', 'execute', '--env-id', ENV_ID,
         '--command', json.dumps(cmd), '--json'],
        capture_output=True, text=True
    )
    if result.returncode != 0:
        return False, record['_id'], result.stdout[:80]
    try:
        out = json.loads(result.stdout)
        if 'error' in out:
            return False, record['_id'], out['error'].get('message', '')
    except:
        pass
    return True, record['_id'], None

def sync_collection(collection, json_file):
    print(f'\n📦 同步集合: {collection}')
    records = []
    with open(json_file, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if line:
                records.append(json.loads(line))
    print(f'   本地数据: {len(records)} 条')

    # 断点续传
    progress_file = os.path.join(SCRIPTS_DIR, f'.progress_{collection}.json')
    done_ids = set()
    if os.path.exists(progress_file):
        with open(progress_file, 'r') as f:
            done_ids = set(json.load(f))
        print(f'   续传: 已完成 {len(done_ids)} 条，剩余 {len(records)-len(done_ids)} 条')

    # 过滤未完成的
    pending = [r for r in records if r.get('_id') not in done_ids]
    print(f'   并发数: {WORKERS}，开始同步...\n')

    success = len(done_ids)
    fail = 0
    total = len(records)

    with ThreadPoolExecutor(max_workers=WORKERS) as executor:
        futures = {executor.submit(upsert_record, collection, r): r for r in pending}
        for i, future in enumerate(as_completed(futures)):
            ok, _id, err = future.result()
            if ok:
                success += 1
                done_ids.add(_id)
            else:
                fail += 1
                print(f'   ❌ {_id}: {err}')

            if (i+1) % 100 == 0:
                # 保存进度
                with open(progress_file, 'w') as f:
                    json.dump(list(done_ids), f)
                print(f'   [{success}/{total}] 成功:{success} 失败:{fail}', flush=True)

    # 完成
    if os.path.exists(progress_file):
        os.remove(progress_file)

    print(f'\n完成: 成功 {success}, 失败 {fail}')
    print('✅ 同步完成')

DATA_FILES = {
    'artworks': os.path.join(SCRIPTS_DIR, 'artworks_data.json'),
    'artists':  os.path.join(SCRIPTS_DIR, 'artists_data.json'),
    'museums':  os.path.join(SCRIPTS_DIR, 'museums_data.json'),
}

def main():
    target = sys.argv[1] if len(sys.argv) > 1 else 'all'
    if target == 'all':
        for col, f in DATA_FILES.items():
            if os.path.exists(f):
                sync_collection(col, f)
    elif target in DATA_FILES:
        sync_collection(target, DATA_FILES[target])
    else:
        print(f'未知: {target}')

if __name__ == '__main__':
    main()

File: src/scripts/sync_to_cloud.js
/**
 * 画说数据同步脚本 - 使用微信云开发HTTP API
 * 用法: node sync_to_cloud.js [collection]
 * 示例: node sync_to_cloud.js artworks
 *       node sync_to_cloud.js artists  
 *       node sync_to_cloud.js museums
 *       node sync_to_cloud.js all
 */

const https = require('https')
const fs = require('fs')
const path = require('path')

const ENV_ID = 'cloudbase-d7gl3kh5vf6b71edc'
const SCRIPTS_DIR = path.dirname(__filename)

// 获取微信云开发访问token
async function getAccessToken() {
  // 从微信开发者工具的本地缓存读取token
  const tokenPath = `${process.env.HOME}/Library/Application Support/微信开发者工具/Default/.access_token`
  const tokenPath2 = `${process.env.HOME}/Library/Application Support/微信开发者工具/WeappSimulator/storage/.access_token`
  
  // 尝试从开发者工具配置读取
  const possiblePaths = [
    tokenPath,
    tokenPath2,
    `${process.env.HOME}/.wxdevtools/access_token`,
  ]
  
  for (const p of possiblePaths) {
    if (fs.existsSync(p)) {
      console.log(`找到token: ${p}`)
      return fs.readFileSync(p, 'utf-8').trim()
    }
  }
  
  return null
}

function httpsRequest(options, body) {
  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = ''
      res.on('data', chunk => data += chunk)
      res.on('end', () => {
        try {
          resolve(JSON.parse(data))
        } catch(e) {
          resolve(data)
        }
      })
    })
    req.on('error', reject)
    if (body) req.write(JSON.stringify(body))
    req.end()
  })
}

async function upsertRecord(token, collection, record) {
  const body = {
    env: ENV_ID,
    query: `db.collection('${collection}').doc('${record._id}').set({data: ${JSON.stringify(record)}})`
  }
  
  const options = {
    hostname: 'api.weixin.qq.com',
    path: `/tcb/databaseupdate?access_token=${token}`,
    method: 'POST',
    headers: { 'Content-Type': 'application/json' }
  }
  
  return httpsRequest(options, body)
}

async function main() {
  console.log('⚠️  此脚本需要微信access_token')
  console.log('请改用以下更简单的方式：\n')
  console.log('📋 增量更新最佳实践：')
  console.log('   新增1条记录时：直接在云开发控制台点"添加记录"')  
  console.log('   更新1条记录时：直接在云开发控制台点击记录编辑')
  console.log('   批量新增时：只导入新增的JSON行（不删除旧数据）\n')
  console.log('✅ 正确的增量导入方法：')
  console.log('   1. 只将新增的记录写入一个临时JSON文件')
  console.log('   2. 云数据库导入时冲突处理选"Insert"（跳过已存在）')
  console.log('   3. 不需要先删除旧数据！')
}

main()
